Intro tidyverse learnr (#8)
EDUCE-UBC/educer: inst/tutorials/tidyverse_ws/tidyverse_ws.Rmd

---
title: "Introduction to the R tidyverse"
author: "Gil J. B. Henriques, Kim Dill-McFarland and Kris Hong.<br>
        A tutorial by [ECOSCOPE](http://ecoscope.ubc.ca/) at UBC."
date: "version `r format(Sys.time(), '%B %d, %Y')`"
description: In-depth description will follow.
output: 
  learnr::tutorial:
      progressive: true
      allow_skip: true
runtime: shiny_prerendered
---

```{r setup, include = FALSE}
# General learnr setup
library(learnr)
knitr::opts_chunk$set(echo = FALSE)
library(educer)
# Helper function to set path to images to "/images" etc.
setup_resources()

# Tutorial specific setup
library(tidyverse) # Suite of packages for data manipulation and visualization
library(lubridate) # Special formatting for dates within data
library(cowplot) # Multi-panel figures
library(readr)

```

## Introduction
In this tutorial, we will learn to manipulate and plot real-world data using the [tidyverse](https://www.tidyverse.org/), a collection of R packages for data analysis and visualization.

You will learn how to:

- Create subsets of rows or columns from data frames using dplyr,
- Change your data frames between wide and long formats,

> - Create various types of graphics,

Reviewer: ```suggestion
- Create various types of data visualizations,
```

- Modify the various features of a graphic, and
- Save your graphic in various formats

This comprehensive introduction to the tidyverse is ideal for beginners but assumes some prior experience with R (e.g. as provided in our "Introduction to R and RStudio" tutorial).

### How to follow this tutorial

```{r child = system.file("resources/markdown", "how_to_follow_tutorials.Rmd", package = "educer")}
```

### Data description

``` {r child = system.file("resources/markdown", "data_description.Rmd", package = "educer")}
```

> During this tutorial, the data are accessible as a data frame called `geochemicals`. In the interactive code box below, use the function `head` to peak at the first few rows of this data frame:

Reviewer: ```suggestion
During this tutorial, the data are accessible as a data frame called `geochemicals`. In the interactive code box below, use the function `head` to peak at the first five rows of this data frame:
```

```{r geochemicals, exercise = TRUE}
head(<name_of_the_data_frame>)
```

```{r geochemicals-solution}
head(geochemicals)
```

### Loading packages

The [tidyverse](https://www.tidyverse.org/) is a collection of R packages. Packages expand R's capabilities by providing additional functions. Some tidyverse packages that we will use in this tutorial are: dplyr (for data manipulation; pronounced *dee-plier*), tidyr (for rearranging data frames), and ggplot2 (for data visualization).  We will also use the packages lubridate (for working with dates) and cowplot (for multi-panel figures).

When you work in your own projects, always load the packages you need at the top of your script. You can load each one of the tidyverse packages individually, but since they are often used together, you can also download and load the entire suite of tidyverse packages at once. In the future, you can use the code below to load all of the R packages we will use in this tutorial. 

> 

Reviewer: ```suggestion
Alternatively, when you are working on a large project with version control, you can create R's equivalent of a virtual environment with the `packrat` package, such that anyone who wishes to build upon your code in the future will do so with the same packages and versions you used while developing the project.
```

Reviewer 2: This introduces two new concepts, reproducibility in general and version control specifically. This does not really address the point the paragraph is making (that you have to load packages even if they are installed IMHO). I would not add this suggestion.

Reviewer: Fair enough! I thought the idea of a `virtualenv` might click for people who are used to Python, but I can see how it is extraneous to this tutorial.

Use the new data frame as an argument for the next step.
The final data frame should have sixteen rows. 

```{r exercise_rename_mutate, exercise = TRUE, exercise.setup = "setup_dat"}

```

```{r exercise_rename_mutate-solution}
selected_data <- select(dat, Date, Depth, Mean_O2)
renamed_data <- rename(selected_data, O2 = Mean_O2)
filtered_data <- filter(renamed_data, month(Date) == 8 & !is.na(O2))
mutate(filtered_data, O2 = O2 * 32)
```

### Piping with `%>%`

In the previous exercise, we had to perform quite a large series of data wrangling tasks.
First, we selected some columns; then, we renamed them; then we filtered them; and finally we mutated one of the columns.
The most natural way to think about this exercise is as a *chain* (or sequence) of functions: each dplyr verb wrangles the input data frame into a new (output) data frame, which is then provided as the input to the following verb.
Metaphorically, the output of the first function is *piped into* the following function.

The tidyverse package magrittr introduced the `%>%` (pipe) operator, which uses this metaphor to make code easier to read:

> `f(x) %>% g(y)` is the same as `g(f(x),y)`

Reviewer: Nice! This is a very succinct way of explaining piping, and this exact line is what made it make sense to me when I was first learning about the pipe operator. Well done!

We can turn data from wide to long, and vice versa, by "pivoting". To turn `dat` into a longer data frame, we use `pivot_longer()`, a function from the tidyverse package tidyr. This function takes the following arguments:

- The data frame to be pivoted;
- A vector containing the columns that we want to combine into a single, longer column **or** the columns that we do *not* want to combine (indicated by a minus sign);
- `names_to`: The name of the column that is created from the data stored in the current column names;
- `values_to`: The name of the column to create from the data stored in cell values.

Here is `pivot_longer` in action:

```{r, pivot_longer, exercise = TRUE, exercise.setup = "setup_dat"}
long_dat  <- dat %>% 
  pivot_longer(-c(Cruise, Date, Depth), 
               names_to = "Chemical", 
               values_to = "Concentration")
long_dat
```

```{r setup_long_dat, include = FALSE}
dat <- select(geochemicals, Cruise, Date, Depth, starts_with("Mean_"))
long_dat <- dat %>% 
  pivot_longer(-c(Cruise, Date, Depth), 
               names_to = "Chemical", 
               values_to = "Concentration")
```

> We can undo this operation with the opposite function, `pivot_wider`:

Reviewer: I know it's similar to `pivot_longer()` but it might be worth explaining what the arguments of `pivot_wider()` actually do as well.

  scale_y_reverse() +
  geom_point() +
  geom_line() +
  facet_wrap(vars(Depth)) 

plot_grid(pA, pB, ncol = 1, labels = c("A", "B"), rel_heights = c(1.5, 1))
```

### Additional resources

In this tutorial, we have gone over just some of the functions and graphics within the tidyverse. Below are some resources for further learning and practice!

- [R cheatsheets](https://www.rstudio.com/resources/cheatsheets/) also available in RStudio under Help > Cheatsheets

- [Introduction to dplyr](https://cran.r-project.org/web/packages/dplyr/vignettes/dplyr.html)
- [dplyr tutorial](https://rpubs.com/justmarkham/dplyr-tutorial)
- [dplyr video tutorial](https://www.r-bloggers.com/hands-on-dplyr-tutorial-for-faster-data-manipulation-in-r/)
- [More functions in dplyr and tidyr](https://rpubs.com/bradleyboehmke/data_wrangling)
- [R for data science (online free book)](https://r4ds.had.co.nz/)

- [ggplot tutorial 1](http://r-statistics.co/Complete-Ggplot2-Tutorial-Part1-With-R-Code.html)
- [ggplot tutorial 2](https://rpubs.com/g_jw/ggplot2_tutorial)
- [ggplot tutorial 3](http://tutorials.iq.harvard.edu/R/Rgraphics/Rgraphics.html#the_1_faq)

> - [ggplot2: Elegant Graphics for Data Analysis (online free book)](https://ggplot2-book.org/)

Reviewer: Looks great to me! Good work, Gil.

setup_resources()

# Tutorial specific setup
library(tidyverse) # Suite of packages for data manipulation and visualization
library(lubridate) # Special formatting for dates within data
library(cowplot) # Multi-panel figures
library(readr)

```

## Introduction
In this tutorial, we will learn to manipulate and plot real-world data using the [tidyverse](https://www.tidyverse.org/), a collection of R packages for data analysis and visualization.

You will learn how to:

- Create subsets of rows or columns from data frames using dplyr,
- Change your data frames between wide and long formats,
- Create various types of data visualizations,
- Modify the various features of a graphic, and
- Save your graphic in various formats

This comprehensive introduction to the tidyverse is ideal for beginners but assumes some prior experience with R (e.g. as provided in our "Introduction to R and RStudio" tutorial).

### How to follow this tutorial

```{r child = system.file("resources/markdown", "how_to_follow_tutorials.Rmd", package = "educer")}
```

### Data description

``` {r child = system.file("resources/markdown", "data_description.Rmd", package = "educer")}
```

> During this tutorial, the data are accessible as a data frame called `geochemicals`. In the interactive code box below, use the function `head` to peak at the first five rows of this data frame:

Reviewer 2: ```suggestion
During this tutorial, the data are accessible as a data frame called `geochemicals`. In the interactive code box below, use the function `head()` to peak at the first five rows of this data frame:
```

## Data wrangling with dplyr

"Wrangling" is the process of manipulating and transforming "raw" data into a more accessible format that you can use for analysis or visualization. 

Subsetting data frames and vectors in base R, like we did in the previous section, is an example of data wrangling. However, base R can quickly become confusing and convoluted. The tidyverse package dplyr provides an alternative syntax for data wrangling. It is so convenient, and integrates so well with other popular tools like ggplot2, that it has rapidly become the de facto standard for data manipulation.

Compared to base R, dplyr code is very readable. Instead of base R’s indexing system (with its brackets and parentheses), which is difficult to read, dplyr operations are based around a small set of functions that match English-language "verbs". The most common verbs are:

- `select()` a subset of columns based on their names.
- `filter()` out a subset of rows based on their values.
- `rename()` columns.
- `arrange()` a column in ascending or descending order.
- `mutate()` all values of a column, i.e., add a new column that is a function of existing columns.

> - `group_by()` a variable and `summarise` data by the grouped variable.

Reviewer 2: ```suggestion
- `group_by()` a variable and `summarise()` data by the grouped variable.
```

dat # print dat
```

```{r dat-hint-2}
dat <- select(geochemicals, Cruise, Date, Depth, starts_with("Mean_")) # create dat
dat # print dat
```

```{r setup_dat, include = FALSE}
dat <- select(geochemicals, Cruise, Date, Depth, starts_with("Mean_"))
```

We will use the `dat` data frame in the rest of this tutorial.

The full list of helper functions can be found by running `?select_helpers` in 
the console. 

### `filter()`

You can use `filter()` to subset particular rows based on a *logical condition* of a variable: instead of choosing (by index) which of the hundreds of rows to keep, we chose to keep only those rows that obey a specific condition: for example, we might want to keep only those observations taken at a depth of 100 m.

To define the logical condition, we use *logical operators*. Here are some examples of logical operators:

R code     | meaning
---------- | ---------------
`==`       | equals
`< or >`   | less/greater than
`<= or >= `| less/greater than or equal to
`%in%`     | in
`is.na`    | is missing (`NA`)
`!`        | not (e.g. not equal to `!=`; not missing `!is.na()`)
`&`        | and

> `\|`        | or

Reviewer 2: ```suggestion
`|`        | or
```

`f(x) %>% g(y)` is the same as `g(f(x),y)`

`select(dat, Cruise)` is the same as `dat %>% select(Cruise)`

Piping works nicely to condense code and to improve readability. It integrates especially well with dplyr. Here is an example:

```{r echo = TRUE}
geochemicals %>%
  select(Cruise, Date, Depth, Mean_CH4) %>%
  filter(Date >= "2008-02-01") %>%
  rename(CH4 = Mean_CH4) %>%
  mutate(log_CH4 = log(CH4)) %>% 
  head()
```

In the code above, we start by selecting some columns and *pipe* the output into the next step (`filter()`). We keep piping the output into the next step of the chain. After we are done wrangling, we pipe the output to the function `head()` (which prints the first few rows of a data frame). Note how dplyr verbs combined with the pipe operator make this complex series of data wrangling operations read almost like an English sentence (think of `%>%` as the word *"then"*: e.g., "select these columns, *then* filter by date, *then* rename this column...")

#### Exercise: Piping

In the box below, rewrite your code from the previous exercise using pipes.

> **Tip:** In RStudio, you can use the keyboard shortcut <kbd>Cmd</kbd>+<kbd>Shift</kbd>+<kbd>m</kbd> to enter the pipe operator.

Reviewer 2: ```suggestion
**Tip:** In RStudio, you can use the keyboard shortcut <kbd>Cmd</kbd>+<kbd>Shift</kbd>+<kbd>M</kbd> to enter the pipe operator.
```

```{r, scatterplot-color, exercise = TRUE,  exercise.setup = "setup_long_dat"}
dat %>%
  ggplot(aes(x = Mean_O2, y = Mean_CO2, colour = Depth)) +
  geom_point()
```

(The `aes()` function accepts both `color` and `colour` as correct argument names.)

As an alternative to point colour, we can also map a continuous variable by using differently sized points. Modify the box below so as to use the `size` aesthetic instead of the `colour` aesthetic:

```{r, scatterplot-color-2, exercise = TRUE,  exercise.setup = "setup_long_dat"}
dat %>%
  ggplot(aes(x = Mean_O2, y = Mean_CO2, colour = Depth)) +
  geom_point()
```

```{r, scatterplot-color-2-solution}
dat %>%
  ggplot(aes(x = Mean_O2, y = Mean_CO2, size = Depth)) +
  geom_point()
```

We can also colour points by a categorical variable. Recall our `long_dat` data frame:

```{r echo=FALSE, message=FALSE, warning=FALSE}
dat <- select(geochemicals, Cruise, Date, Depth, starts_with("Mean_"))
long_dat <- dat %>% 
  pivot_longer(-c(Cruise, Date, Depth), 
               names_to = "Chemical", 
               values_to = "Concentration")
long_dat
```

> In the box below, `filter` this data frame so that it includes only oxygen and carbon dioxide. Then, plot the relationship between `Concentration` (in the x-axis) and `Depth` (in the y-axis), indicating the `Chemical` variable by point color:

Reviewer 2: ```suggestion
In the box below, `filter()` this data frame so that it includes only oxygen and carbon dioxide. Then, plot the relationship between `Concentration` (in the x-axis) and `Depth` (in the y-axis), indicating the `Chemical` variable by point color:
```

```{r, scatterplot-facet-grid, exercise = TRUE,  exercise.setup = "setup_long_dat"}
long_dat %>%
  filter(Cruise %in% 10:13, 
         Chemical %in% c("Mean_O2", "Mean_CO2")) %>% 
  ggplot(aes(x = Concentration, y = Depth)) +
  geom_point() +
  scale_y_reverse() +
  <facet by Cruise and Chemical>
```

```{r, scatterplot-facet-grid-solution}
long_dat %>%
  filter(Cruise %in% 10:13, 
         Chemical %in% c("Mean_O2", "Mean_CO2")) %>% 
  ggplot(aes(x = Concentration, y = Depth)) +
  geom_point() +
  scale_y_reverse() +
  facet_grid(vars(Cruise), vars(Chemical))
```

### Fine-tuning your plot

So far, we have produced plots for data *exploration*. We didn't care too much about how they looked. But when we create plots for *communication* (i.e., for slideshows, posters, or scientific articles) we need to make sure that they also look good. 

In this section, we will work to create a beautiful multi-panel figure. Each panel will be based on a plot that we produced so far. Panel A will show the negative correlation between O$_2$ and CO$_2$ concentration, and panel B will show paired boxplots for CO$_2$ and O$_2$ above and below 100 meters. 

#### Labels

> We can change the labels of the previous plot with the `labs` function, specifying exactly what we want as the title, x-, and y-axis labels:

Reviewer 2: ```suggestion
We can change the labels of the previous plot with the `labs()` function, specifying exactly what we want as the title, x-, and y-axis labels:
```

               values_to = "Concentration")
```

We can undo this operation with the opposite function, `pivot_wider()`, which works similarly to its counterpart. The `pivot_wider()` function takes the following arguments:

- The data frame to be pivoted;

> - A vector containing the columns that we want to pivot **or** the columns that we do *not* want to pivot (indicated by a minus sign). If this argument is left blank, it defaults to all columns in the data frame except for those specified in `names_from` and `values_from`;

Reviewer 2: ```suggestion
- A vector containing the columns that we want to pivot **or** the columns that we do *not* want to pivot (indicated by an exclamation point `!`). If this argument is left blank, it defaults to all columns in the data frame except for those specified in `names_from` and `values_from`;
```
You new syntax now uses `!` for the anti-(not)-set.
